src/Compiler/Interpreter/VMCode.idr

```diff
@@ -0,0 +1,281 @@
+module Compiler.Interpreter.VMCode
+
+
+import Core.Core
+import Core.Context
+import Core.Context.Log
+import Core.Primitives
+import Core.Value
+import Compiler.Common
+import Compiler.VMCode
+import Data.IOArray
+import Libraries.Data.NameMap
+import Data.Nat
+import Data.Vect
+
+public export
+data Object : Type where
+    Closure : (predMissing : Nat) -> (args : List Object) -> Name -> Object
+    Constructor : (tag : Either Int Name) -> (args : List Object) -> Object
+    Const : Constant -> Object
+    Null : Object
+
+showType : Object -> String
+showType (Closure _ _ _) = "Closure"
+showType (Constructor _ _) = "Constructor"
+showType (Const _) = "Constant"
+showType Null = "Null"
+
+mutual
+    showSep : Nat -> List Object -> String
+    showSep k [] = ""
+    showSep k [o] = showDepth k o
+    showSep k (o :: os) = showDepth k o ++ ", " ++ showSep k os
+
+    showDepth : Nat -> Object -> String
+    showDepth (S k) (Closure mis args fn) = show fn ++ "-" ++ show mis ++ "(" ++ showSep k args ++ ")"
+    showDepth (S k) (Constructor (Left t) args) = "tag" ++ show t ++ "(" ++ showSep k args ++ ")"
+    showDepth (S k) (Const c) = show c
+    showDepth _ obj = showType obj
+
+Show Object where
+    show = showDepth 5
+
+data State : Type where
+record InterpState where
+    constructor MkInterpState
+    defs : NameMap VMDef
+    locals : IOArray Object
+    return : Maybe Object
```

Comment:
I think `return` is a bit confusing here. Later down there's something that looks like:
```
case return <$> get State of
  Nothing => branch1
  Just obj => branch2
```
and my first reaction was to suggest regactoring it as
```
obj <- get State
branch2
```
because it looked like `return <$>` was just going under the `Core` layer and adding
a `Just` constructor! It took me quite some time to figure out that the code was fine
as is.

Reply:
Ah, ok. I never ever used `return` instead of `pure` when I used Haskell, but I'll change it to avoid confusion.